File: Sunrin-ICPC-2022/Pre-E/PE-slow.py
n = int(input())
a = [0] + list(map(int, input().split()))

c = [0] * 101
idx = [[0 for col in range(101)] for row in range(n+1)]
lst = [0] * 101

for i in a: c[i] += 1
mx = max(c)

if n >= 100 * 100:
    print(mx)
else:
    for i in range(1, 101): lst[i] = -1
    for i in range(n, 0, -1):
        for j in range(1, 101): idx[i][j] = lst[j]
        lst[a[i]] = i
    for d in range(-100, 101):
        if d == 0: continue
        for i in range(1, n):
            j, len = i, 0
            while j != -1:
                nxt = a[j] + d
                len += 1
                if 1 <= nxt <= 100: j = idx[j][nxt]
                else: break
            mx = max(mx, len)
    print(mx)

File: Sunrin-ICPC-2022/Pre-E/PE-slow.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().split('\n');

const n = parseInt(input[0]);
let a = input[1].split(' ').map(s => parseInt(s));
a.unshift(0);

let count = new Array(101).fill(0);
let idx = new Array(n+1).fill(0).map(_ => new Array(101).fill(0));
let lst = new Array(101).fill(0);
let mx = 0;

for(let i=1; i<=n; i++) count[a[i]] += 1;
for(let i=1; i<=100; i++) mx = Math.max(mx, count[i]);

for(let i=1; i<=100; i++) lst[i] = -1;
for(let i=n; i>=1; i--){
    for(let j=1; j<=100; j++) idx[i][j] = lst[j];
    lst[a[i]] = i;
}

for(let d=-100; d<=100; d++){
    if(d == 0) continue;
    for(let i=1; i<=n; i++){
        let len = 0;
        for(let j=i; j!=-1; ){
            let nxt = a[j] + d; len += 1;
            if(1 <= nxt && nxt <= 100) j = idx[j][nxt];
            else break;
        }
        mx = Math.max(mx, len);
    }
}

console.log(mx);